### PR TITLE
fix: survey editor validation

### PR DESCRIPTION
**Pull request by pandeymangg**
## What does this PR do?

Fixes survey editor validation with zod refinements as the center-piece

Fixes # (issue)

## How should this be tested?

- Test A
- Test B

### Required

- [ ] Filled out the "How to test" section in this PR
- [ ] Read [How we Code at Formbricks](<[https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md](https://formbricks.com/docs/contributing/how-we-code)>)
- [ ] Self-reviewed my own code
- [ ] Commented on my code in hard-to-understand bits
- [ ] Ran `pnpm build`
- [ ] Checked for warnings, there are none
- [ ] Removed all `console.logs`
- [ ] Merged the latest changes from main onto my branch with `git pull origin main`
- [ ] My changes don't cause any responsiveness issues
- [ ] First PR at Formbricks? [Please sign the CLA!](https://cla-assistant.io/formbricks/formbricks) Without it we wont be able to merge it 🙏

### Appreciated

- [ ] If a UI change was made: Added a screen recording or screenshots to this PR
- [ ] Updated the Formbricks Docs if changes were necessary

### CODE REVIEW

File: packages/types/surveys/types.ts [hunk 1]
```diff
@@ -489,48 +547,381 @@ export const ZSurveyInlineTriggers = z.object({
...
+        "lowerLabel",
+        "label",
+        "placeholder",
+        "backButtonLabel",
+      ];
+
+      const fieldsToValidate =
+        questionIndex === 0
+          ? initialFieldsToValidate.filter((_, idx) => idx !== initialFieldsToValidate.length - 1)
+          : initialFieldsToValidate;
```
Comment (gupta-piyush19) [inline packages/types/surveys/types.ts:718]: can you please simplify this piece of code with
``` ts
    const initialFieldsToValidate = [
        "html",
        "buttonLabel",
        "upperLabel",
        "lowerLabel",
        "label",
        "placeholder",
      ];

      const fieldsToValidate =
        questionIndex === 0 ? initialFieldsToValidate : [...initialFieldsToValidate, "backButtonLabel"];```

File: packages/types/surveys/types.ts [hunk 2]
```diff
@@ -489,48 +547,381 @@ export const ZSurveyInlineTriggers = z.object({
...
+    segment: ZSegment.nullable(),
+    singleUse: ZSurveySingleUse.nullable(),
+    verifyEmail: ZSurveyVerifyEmail.nullable(),
+    pin: z.string().min(4, { message: "PIN must be a four digit number" }).nullish(),
+    resultShareKey: z.string().nullable(),
+    displayPercentage: z.number().min(0.01).max(100).nullable(),
+    languages: z.array(ZSurveyLanguage),
   })
-);
+  .superRefine((survey, ctx) => {
```
Comment (gupta-piyush19) [inline packages/types/surveys/types.ts:598]: Please add a few comments in the superRefine handle where you think clarification is necessary.

**Review by gupta-piyush19 — CHANGES_REQUESTED** [review]
Thanks, @pandeymangg, for the PR. 🙌 
Survey editor works great, and the changes look good.
I've added a few comments. Can you please check them out? 🤓 
I'm still testing  and will add more comments in case of any issues in the functionality

**Comment by gupta-piyush19** [comment]
A few more comments:
1. Please fix build errors, unit and e2e tests.
2.  image I can save the website survey without selecting any survey trigger.
3. image - Getting generic error in case of empty inputs, can we add the config name to add more context in the error message?
4. image - casing looks a bit off, take help from @jobenjada to find out the best error message 🤓 
Thanks 🙏

**Comment by gupta-piyush19** [comment]
Thanks @pandeymangg, for the changes 🙌 
LGTM 🚀 
@jobenjada, Can you please test the UI/UX of the survey editor and suggest some error messages? 🤓

**Review by gupta-piyush19 — APPROVED** [review]

**Review by jobenjada — CHANGES_REQUESTED** [review]
Hey Pandeyman,

this looks really good, thanks for your work!! 💪 

Regarding UX I have a few points:

1. Align field names with UI labels: There is a mismatch between the name of the field and the label of the input field:

subheader vs. Description
image

headline vs. Question 
image

image

image


---------

2. Capitalize the first word in all error messages: If error messages start with a variable, they're not capitalized. I found a work around by adding "The" which I also find to be more user-friendly:

image

I'm not sure if I have caught all cases with all error messages, please double check an either add an article or capitalize the variable.

--------

4. The error messages are written from a code perspective, not form a user perspective. For example: "The headline in question 1 is not valid for the following languages: Abkhazian"

What does "not valid" mean? I see that there are two cases: "invalid" and "not valid for languages"

Does it just mean they are missing? Ideally, we'd reframe the error message to an information the user can act on. If we say its "missing:, they know they have to add it. If we say its "invalid" the user doesn't know how to validate it. 

------

5. Cyclic Logic: Is logic which jumps to a question ahead of the question where the logic is set always cyclic? If so, why don't we filter out the questions from the select list?

------

Not sure if I've found all of the error messages ;) ping me if I missed some 

Thanks! 🤙

**Review by mattinannt — APPROVED** [review]
@pandeymangg Works great!! 💪
Thanks a lot for this revamp; it really improves the survey editor and makes sure we validate everything and can point to what information is missing or in the wrong format 💪🚀

**Pull request merged**